broadinstitute/single_cell_portal_core
# Image Pipeline: improve robustness by using workspace buckets (SCP-4562)

This improves Image Pipeline robustness and enables it to run end-to-end as a standalone component.

### Technical summary
Previously, robustness suffered because PAPI runs required fetching data from the staging SCP API.  Now, PAPI runs can read data from cached files in the workspace bucket.  These files are smaller than comparable staging API responses.  Between the smaller size and the simpler, separate server-side solution (i.e., buckets), Image Pipeline can now generate extremely many images faster and more scalable.

Image Pipeline can now produce over 8k gene expression scatter plot images in a run.  That's > 7x more than before.  Images can now be written to the SCP study's Terra workspace bucket, rather than a hard-coded test bucket.

### Manual test
1. Ensure you can read from Vault:
```
vault login -method=github token=\`~/bin/git-vault-token\`
```

2. Authenticate gcloud on terminal: `gcloud auth login`

3. Confirm you're auth'd with your Broad account.  (Your Broad username should display instead of "eweitz".): 
```
gcloud auth list | grep "*"
*       eweitz@broadinstitute.org
```

4. Ensure you're on the staging GCP project
```
gcloud config set project broad-singlecellportal-staging
```

5.  Pull secrets from Vault:
> `vault read secret/kdux/scp/staging/scp_service_account.json -format=json | jq .data > /tmp/scp_service_account_staging.json; vault read secret/kdux/scp/staging/scp_config.json -format=json | jq .data > /tmp/scp_config_staging.json`

6. Make secrets accessible as environment variables:
> `SA_EMAIL_ADDRESS=$(cat /tmp/scp_service_account_staging.json | jq -r .client_email); GCP_NETWORK=$(cat /tmp/scp_config_staging.json | jq -r .GCP_NETWORK_NAME); STAGING_INTERNAL_IP=$(cat /tmp/scp_config_staging.json | jq -r .APP_INTERNAL_IP);`

7. Run Image Pipeline on staging (new: `--bucket` and `--cluster`, version based on 0.1.0 instead of 0.0.1):
> `gcloud alpha genomics pipelines run --regions us-central1 --env-vars="STAGING_INTERNAL_IP=$STAGING_INTERNAL_IP" --command-line 'IS_PAPI=1 NODE_TLS_REJECT_UNAUTHORIZED=0 node expression-scatter-plots.js --accession="SCP303" --bucket="fc-37c32921-4c79-4767-819d-66f157f871a9" --cluster="All_Cells_UMAP" --environment="staging" --debug-headless; # human milk DE pilot' --docker-image gcr.io/broad-singlecellportal-staging/image-pipeline:0.1.0_$COMMIT_ID --service-account-email $SA_EMAIL_ADDRESS --network $GCP_NETWORK --subnetwork $GCP_NETWORK # Command to run Image Pipeline on PAPI`

8. Manually copy your run's operation ID from above output, e.g. "15137633979837466304" in below:
```
Running [projects/broad-singlecellportal-staging/operations/15137633979837466304].
```

9. Watch for your PAPI run to complete: replace "15137633979837466304" with your value from previous step
> `OPERATION_ID=15137633979837466304; gcloud alpha genomics operations wait ${OPERATION_ID}; tput bel; gcloud alpha genomics operations describe ${OPERATION_ID} | grep exitStatus; tput bel`

10. Confirm your exit status code is 0, e.g.:
```
Waiting for [projects/broad-singlecellportal-staging/operations/15137633979837466304]...done.                                                                                                                                                
      exitStatus: 0
```

### Next steps
Until Image Pipeline reliably caches all genes, we can prioritize which genes get a cached plot image.  Details: SCP-4789.

This satisfies SCP-4562.
